### PR TITLE
Rewrite TensorFlow train_step and test_step

### PR DESCRIPTION
**Pull request by Rocketknight1**
Draft PR for a full rewrite of the TF train/test steps. I swear this will fix like 50% of our TF issues in one PR.

Current status:
- Correctly handles output mapping across most model classes for losses + metrics
- Keras metrics are back, even with the dummy loss. (!!!!)
- Keras metrics work correctly even for multi-output models (like QA)
- In most cases, users can pass tensors in either the input dict or the labels and the model will handle them correctly.
- No more errors when calling `fit()` when the model has nested output structure (e.g. the model outputting a `past` tuple)

What's left to do:

- [X] Models with multiple unusual outputs that do not match label names may still have issues with metrics. This is relatively uncommon. We support adding a property to those classes to tell Keras what to do with the labels, but we haven't added it to any models yet. (None are failing in tests, so hopefully we won't need to worry too much about this!)
- [x] Testing testing testing! I want to rerun all notebooks/examples and make sure the user experience is good.
- [X] CI testing - We need to make sure we don't regress on any of this
- [ ] Discoverability: After this is merged we should update notebooks/examples to show off the cool new features, and document our TF workflow/philosophy somewhere that new users will find.

### CODE REVIEW

**Comment by Rocketknight1** [comment]
(Requesting reviews now that @gante is back)

File: src/transformers/modeling_tf_utils.py [hunk 1]
```diff
@@ -990,22 +996,54 @@ def train_step(self, data):
...
                     if key in arg_names and key not in x:
                         x[key] = val
+                    elif output_to_label.get(key, None) in arg_names and key not in x:
+                        x[output_to_label[key]] = val
+        if y is None:
+            y = {key: val for key, val in x.items() if key in label_kwargs}
+            if not y and not self._using_dummy_loss:
+                raise ValueError("Could not find label column(s) in input dict and no separate labels were provided!")
+            if len(y) == 1:
+                _, y = y.popitem()
```
Comment (gante) [inline src/transformers/modeling_tf_utils.py:1006]: This converts `y` from a dictionary with one item to the value of that dictionary entry. Looking below, it seems like it should handle dicts correctly. What's happening here?

**Review by gante — COMMENTED** [review]
<3 This is great, Keras users will definitely feel more at home

I've added two comments: a suggestion (for potentially more organized code) and a question. Other than that, LGTM!

File: src/transformers/modeling_tf_utils.py [hunk 2]
```diff
@@ -957,25 +952,36 @@ def compute_loss(self, *args, **kwargs):
...
+        if self._label_to_output_map is not None:
+            label_to_output = self._label_to_output_map
+        elif "start_positions" in arg_names:
+            label_to_output = {"start_positions": "start_logits", "end_positions": "end_logits"}
+        elif "sentence_order_label" in arg_names:
+            label_to_output = {"labels": "prediction_logits", "sentence_order_label": "sop_logits"}
+        elif "next_sentence_label" in arg_names:
+            label_to_output = {"labels": "prediction_logits", "next_sentence_label": "seq_relationship_logits"}
+        elif "mc_labels" in arg_names:
+            label_to_output = {"labels": "logits", "mc_labels": "mc_logits"}
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py:973]: I explored this when writing the PR! I think that would work in a lot of cases, but there are some models which have their own custom losses, and other models that define `hf_compute_loss` in the model class itself.

So I'm not sure if moving this to the `Loss` classes would be that easy, but for cleanliness, I can extract this to a method called something like `infer_label_to_output_map()` and just call that in `train_step` instead?

File: src/transformers/modeling_tf_utils.py (hunk 1 above)
Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py]: Fixed! This code was added because the user often passes a dict where the key is "labels", which is not the name of any of the outputs. The correct thing to do for those models is to map the "labels" tensor to the first model output - I changed this line so that it checks the single key is called "labels" before doing so.

Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py]: Fixed it up a little more - now we try to map by key name before falling back to mapping to the first output as a last resort.

File: src/transformers/modeling_tf_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/modeling_tf_utils.py:973]: Extracting to an external function sounds good 👍  (especially because it is reused between train and test)

**Review by gante — APPROVED** [review]
LGTM 👍
